File: ruby/name_TEST.rb
#!/usr/bin/ruby
# vim: set sw=4 sts=4 et tw=80 :

#
# Copyright (c) 2006 Ciaran McCreesh <ciaranm@ciaranm.org>
#
# This file is part of the Paludis package manager. Paludis is free software;
# you can redistribute it and/or modify it under the terms of the GNU General
# Public License version 2, as published by the Free Software Foundation.
#
# Paludis is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA
#

require 'test/unit'
require 'Paludis'

module NameTestCaseBase
    def valid_name_foo
        return "foo"
    end

    def valid_name_bar
        return "bar"
    end

    def bad_name
        return "foo~"
    end

    def test_create
        x = name_type().new(valid_name_foo())
    end

    def test_assign
        x = name_type().new(valid_name_foo())
        y = x
    end

    def test_compare
        x = name_type().new(valid_name_foo())
        y = name_type().new(valid_name_foo())
        z = name_type().new(valid_name_bar())
        assert_equal x, y
        assert_not_equal y, z
        assert_operator z, :<, x
        assert_operator x, :>, z
        assert_raise TypeError do
            x < valid_name_foo()
        end
    end

    def test_create_errors
        assert_raise error_type() do
            x = name_type().new(bad_name())
        end
        assert_raise error_type() do
            x = name_type().new("")
        end
    end

    def test_to_s
        x = name_type().new(valid_name_foo())
        y = name_type().new(valid_name_bar())
        assert_equal valid_name_foo(), x.to_s
        assert_equal valid_name_bar(), y.to_s
    end
end

class Paludis
    class TestCase_PackageNamePart < Test::Unit::TestCase
        include NameTestCaseBase

        def error_type
            return PackageNamePartError
        end

        def name_type
            return PackageNamePart
        end
    end

    class TestCase_CategoryNamePart < Test::Unit::TestCase
        include NameTestCaseBase

        def error_type
            return CategoryNamePartError
        end

        def name_type
            return CategoryNamePart
        end

        def test_plus
            q = CategoryNamePart.new("foo") + PackageNamePart.new("bar")
            assert_equal "foo/bar", q.to_s
        end

        def test_plus_bad
            assert_raise TypeError do
                q = CategoryNamePart.new("foo") + "bar"
            end
        end
    end

    class TestCase_UseFlagName < Test::Unit::TestCase
        include NameTestCaseBase

        def error_type
            return UseFlagNameError
        end

        def name_type
            return UseFlagName
        end
    end

    class TestCase_RepositoryName < Test::Unit::TestCase
        include NameTestCaseBase

        def error_type
            return RepositoryNameError
        end

        def name_type
            return RepositoryName
        end
    end

    class TestCase_SlotName < Test::Unit::TestCase
        include NameTestCaseBase

        def error_type
            return SlotNameError
        end

        def name_type
            return SlotName
        end
    end

    class TestCase_KeywordName < Test::Unit::TestCase
        include NameTestCaseBase

        def error_type
            return KeywordNameError
        end

        def name_type
            return KeywordName
        end
    end

    class TestCase_QualifiedPackageName < Test::Unit::TestCase
        include NameTestCaseBase

        def error_type
            return QualifiedPackageNameError
        end

        def name_type
            return QualifiedPackageName
        end

        def valid_name_foo
            return "foo/foo"
        end

        def valid_name_bar
            return "bar/bar"
        end

        def test_two_arg_create
            x = QualifiedPackageName.new(CategoryNamePart.new("foo"), PackageNamePart.new("bar"))
            assert_equal "foo/bar", x.to_s
        end

        def test_bad_arg_count_create
            assert_raise ArgumentError do
                x = QualifiedPackageName.new("foo", "bar", "baz")
            end
        end

        def test_bad_type_create
            assert_raise TypeError do
                x = QualifiedPackageName.new("foo", "bar")
            end
        end
    end
end

File: python/name_TEST.py
#!/usr/bin/env python
# vim: set fileencoding=utf-8 sw=4 sts=4 et :

#
# Copyright (c) 2007 Piotr Jaroszyński
#
# This file is part of the Paludis package manager. Paludis is free software;
# you can redistribute it and/or modify it under the terms of the GNU General
# Public License version 2, as published by the Free Software Foundation.
#
# Paludis is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA
#

from paludis import *
import unittest


class TestCase_Names(unittest.TestCase):
    def test_1_create(self):
        self.names = {}
        self.names["cat-foo"] = CategoryNamePart("cat-foo")
        self.names["pkg"] = PackageNamePart("pkg")
        self.names["cat-foo/pkg"] = QualifiedPackageName(
            self.names["cat-foo"], self.names["pkg"]
        )
        self.names["cat-blah/pkg"] = QualifiedPackageName("cat-blah/pkg")
        self.names["3.3"] = SlotName("3.3")
        self.names["repo"] = RepositoryName("repo")
        self.names["keyword"] = KeywordName("keyword")
        self.names["set"] = SetName("set")

    def test_2_create_error(self):
        self.assertRaises(PackageNamePartError, PackageNamePart, ":bad")
        self.assertRaises(CategoryNamePartError, CategoryNamePart, ":bad")
        self.assertRaises(CategoryNamePartError, QualifiedPackageName, ":bad")
        self.assertRaises(SlotNameError, SlotName, ":bad")
        self.assertRaises(RepositoryNameError, RepositoryName, ":bad")
        self.assertRaises(KeywordNameError, KeywordName, ":bad")
        self.assertRaises(SetNameError, SetName, ":bad")
        self.assertRaises(Exception, PackageNamePartIterable)
        self.assertRaises(Exception, CategoryNamePartIterable)
        self.assertRaises(Exception, QualifiedPackageNameIterable)

    def test_3_str(self):
        self.test_1_create()
        for (k, v) in self.names.items():
            self.assertEqual(str(v), k)

    def test_4_operators(self):
        self.assert_(
            CategoryNamePart("cat-foo") + PackageNamePart("pkg")
            == QualifiedPackageName("cat-foo/pkg")
        )

    def test_5_data_members(self):
        qpn = QualifiedPackageName("cat/foo")
        self.assertEqual(str(qpn.category), "cat")
        self.assertEqual(str(qpn.package), "foo")


if __name__ == "__main__":
    unittest.main()
